src/main/native/jni_utils.cc

```diff
@@ -21,38 +20,130 @@ JNIEnv *getJNIEnv() {
   return pEnv;
 }
 
+// Class fully qualified names
+const char *CLS_NODE = "org/bblfsh/client/v2/Node";
+const char *CLS_CTX = "org/bblfsh/client/v2/Context";
+const char *CLS_OBJ = "java/lang/Object";
+const char *CLS_RE = "java/lang/RuntimeException";
+
+const char *CLS_JNODE = "org/bblfsh/client/v2/JNode";
+const char *CLS_JNULL = "org/bblfsh/client/v2/JNull";
+const char *CLS_JSTR = "org/bblfsh/client/v2/JString";
+const char *CLS_JINT = "org/bblfsh/client/v2/JInt";
+const char *CLS_JFLT = "org/bblfsh/client/v2/JFloat";
+const char *CLS_JBOL = "org/bblfsh/client/v2/JBool";
+const char *CLS_JUINT = "org/bblfsh/client/v2/JUint";
+const char *CLS_JARR = "org/bblfsh/client/v2/JArray";
+const char *CLS_JOBJ = "org/bblfsh/client/v2/JObject";
+
+// Method signatures
+const char *METHOD_JNODE_KEY_AT = "(I)Ljava/lang/String;";
+const char *METHOD_JNODE_VALUE_AT = "(I)Lorg/bblfsh/client/v2/JNode;";
+const char *METHOD_JOBJ_ADD =
+    "(Ljava/lang/String;Lorg/bblfsh/client/v2/JNode;)Lscala/collection/"
+    "mutable/Buffer;";
+const char *METHOD_JARR_ADD =
+    "(Lorg/bblfsh/client/v2/JNode;)Lscala/collection/mutable/Buffer;";
+
+const char *METHOD_OBJ_TO_STR = "()Ljava/lang/String;";
+
+void checkJvmException(std::string msg) {
+  JNIEnv *env = getJNIEnv();
+  if (env->ExceptionCheck()) {
+    // env->ExceptionDescribe(); // prints to stdout, un-comment for debuging
+    auto err = env->ExceptionOccurred();
+
+    jmethodID toString = env->GetMethodID(env->FindClass(CLS_OBJ), "toString",
+                                          METHOD_OBJ_TO_STR);
+    jstring s = (jstring)env->CallObjectMethod(err, toString);
+    const char *utf = env->GetStringUTFChars(s, 0);
+    env->ReleaseStringUTFChars(s, utf);
+
+    env->ExceptionClear();
+
+    auto exception = env->FindClass(CLS_RE);
+    env->ThrowNew(exception, msg.append(": ").append(utf).data());
+  }
+}
+
 jobject NewJavaObject(JNIEnv *env, const char *className, const char *initSign,
                       ...) {
   jclass cls = env->FindClass(className);
-  if (env->ExceptionOccurred() || !cls) {
-    return NULL;
-  }
+  checkJvmException(std::string("failed to find a class ").append(className));
 
   jmethodID initId = env->GetMethodID(cls, "<init>", initSign);
-  if (env->ExceptionOccurred() || !initId) {
-    return NULL;
-  }
+  checkJvmException(
+      std::string("failed to call <inti> constructor for ").append(className));
 
   va_list varargs;
   va_start(varargs, initSign);
   jobject instance = env->NewObjectV(cls, initId, varargs);
   va_end(varargs);
-  if (env->ExceptionOccurred() || !instance) {
-    return NULL;
-  }
+  checkJvmException(
+      std::string("failed get varargs for constructor of ").append(className));
 
   return instance;
 }
 
 jfieldID getField(JNIEnv *env, jobject obj, const char *name) {
   jclass cls = env->GetObjectClass(obj);
-  if (env->ExceptionOccurred() || !cls) {
-    return nullptr;
-  }
+  checkJvmException("failed get an object class");
```

Comment:
Include the class name here too?

Reply:
We do not have a name of the class at this point.